README.md
# alda-clj

[![Clojars Project](https://img.shields.io/clojars/v/io.djy/alda-clj.svg)](https://clojars.org/io.djy/alda-clj)

[![cljdoc badge](https://cljdoc.org/badge/io.djy/alda-clj)](https://cljdoc.org/d/io.djy/alda-clj/CURRENT)

alda-clj is a Clojure library for algorithmic music composition and live-coding
with [Alda](https://alda.io).

A simple Clojure DSL provides useful functions like `note` and `chord` that can
be composed together to create a musical score. The resulting values are
translated into a string of Alda code and played in a subprocess via the `alda`
command line client.

alda-clj is intended to be used in at least two ways:

1. In an **improv / live-coding** setting, using a live Clojure REPL to generate
   and play music on the fly.

2. In a **music composition** setting, where the composition is a Clojure
   program / script that uses alda-clj to generate a score.

## Usage

> **NOTE**: alda-clj requires Alda version 2.0.0 or greater.

1. If you haven't already, [install Alda](https://alda.io/install) and make sure
   `alda` is available on your `PATH`.

  > alda-clj will shell out and use `alda` (wherever it's found on your `PATH`) to
  > play your scores. If desired, you can specify an alternate `alda` executable
  > by binding `alda.core/*alda-executable*` to something else, e.g.
  > `"/home/dave/Downloads/some-other-alda"`.

2. Add the latest release version of alda-clj to your dependencies:

  ```clojure
  ;; deps.edn
  io.djy/alda-clj {:mvn/version "X.X.X"}

  ;; lein/boot
  [io.djy/alda-clj "X.X.X"]
  ```

3. Require `alda.core` and you're off to the races!

  ```clojure
  (require '[alda.core :refer :all])

  (play!
    (part "piano")
    (for [notes [[:c :e :g] [:c :f :a] [:c :e :g]]]
      (apply chord (map #(note (pitch %)) notes))))
  ```

## Docs, examples, etc.

[API documentation](https://cljdoc.org/d/io.djy/alda-clj/CURRENT/api/alda), a
[Getting Started](https://cljdoc.org/d/io.djy/alda-clj/CURRENT/doc/getting-started) guide and
more are available at cljdoc.

There are also [example scripts](examples) in this repo that will give you a
sense of what you can do with alda-clj.

Ping `@dave` on [Slack](https://slack.alda.io) if you have any questions or if
you just want to chat about alda-clj!

## License

Copyright © 2018-2021 Dave Yarwood

Distributed under the Eclipse Public License version 2.0.


CHANGELOG.md
# CHANGELOG

## 0.3.3 (2021-07-24)

(No changes to the library itself.)

Fixed another pom.xml-related issue that was preventing cljdoc from generating
the online docs.

## 0.3.2 (2021-07-24)

(No changes to the library itself.)

Fixed another pom.xml-related issue that was preventing cljdoc from generating
the online docs.

## 0.3.1 (2021-07-24)

Fixed a build issue where the deployed version of alda-clj 0.3.0 wasn't
including the right dependencies because of a stale pom.xml.

I am now generating the pom.xml every time before deploying to Clojars, so this
issue hopefully won't happen again!

## 0.3.0 (2021-06-30)

This release coincides with the release of Alda 2.0.0. :tada:

### Breaking changes

* **alda-clj now requires Alda version 2.0.0 or greater.**

* The syntax for `key-signature` and friends (`key-sig`, `key-sig!`,
  `key-signature!`) has changed slightly, corresponding to the same change in
  Alda itself in Alda 2. See the [Alda 2 migration guide][alda-2-syntax-changed]
  for more information about this change. To summarize:

  ```clojure
  ;; Alda 1 and 2 (still works)
  (key-sig "f+ c+ g+")

  ;; Alda 1 (no longer works in alda-clj)
  (key-sig! [:a :major])
  (key-sig! {:f [:sharp] :c [:sharp] :g [:sharp]})

  ;; Alda 2
  (key-sig '(a major))
  (key-sig '(f (sharp) c (sharp) g (sharp)))
  ```

  Everything else about the syntax in Alda 2 and alda-clj remains the same.

* Removed `*alda-history*` and the `clear-history` function, as they are no
  longer relevant. See below about connecting to an Alda REPL server to preserve
  context between evaluations.

### Improvements

* The `alda` function (as well as the functions that depend on it, including
  `play!`) now prints stdout and stderr output from Alda as it is produced. The
  stdout output is still returned at the end as a string.

### New features

* You can connect to an Alda REPL server (a new feature of Alda 2) via the new
  `connect!` function. There is also a `disconnect!` if you want to undo that
  and revert to the default, out-of-the box behavior, where `alda play` is used
  to evaluate and play scores in a separate context.

  Connecting to an Alda REPL server is recommended, as it allows you to evaluate
  subsequent snippets of Alda code within the same context. This is great for
  interactive development and live coding. I've updated the [Getting
  Started][getting-started] guide with more details about this new workflow, so
  check that out!

* Added `score-text` and `new-score!` functions to display the current score
  text and start a new score, respectively. These functions can only be used
  when connected to an Alda REPL server.

* Added a `send-nrepl-message!` function that most of you won't ever need to
  use, but if you ever do, it's there! `send-nrepl-message!` is used internally
  by functions like `play!` when you're connected to an Alda REPL server.

* Added a `parse-events` function that takes any combination of strings of Alda
  code and alda-clj event records, runs them through `alda parse --output
  events`, and returns a sequence of alda-clj event records that results from
  translating the JSON output.

  Thanks, [@wcerfgba][wcerfgba], for the contribution!

## 0.2.3 (2020-01-27)

No change; bumped version to publish improvements to the Getting Started docs.

## 0.2.2 (2019-08-17)

* Implemented the LispForm protocol on Milliseconds note duration objects. This
  fixes a bug where `midi-note` and `ms` couldn't be used together, e.g.
  `(note (midi-note 60) (ms 2000))`.

## 0.2.1 (2019-08-17)

* Added `midi-note` function that can be used to specify the pitch of a note as
  a MIDI note number, instead of `pitch` which specifies pitch as a letter and
  accidentals.

* Implementation detail: added a LispForm protocol and `->lisp-form` wrapper
  function. Objects that implement the LispForm protocol are representable in
  alda-lisp, and `->lisp-form` returns the corresponding Lisp S-expression for
  that object.

  These S-expressions are rendered directly into Alda as opposed to having a
  syntax representation like `c+`. This allows us to make use of Alda features
  that are implemented only in Lisp to avoid adding additional syntax to the
  language.

  Example: `(->lisp-form (note (midi-note 42))) => (note (midi-note 42))`

## 0.1.8 (2019-02-21)

No change; bumped version to tweak the docs.

## 0.1.7 (2019-01-21)

No change; bumped version to make more tweaks/fixes to the docs.

## 0.1.6 (2019-01-21)

No change; bumped version to make some tweaks to documentation.

## 0.1.5 (2019-01-21)

* Bugfix: `cram` duration can be `nil` (i.e. the default duration) now.

* Emitted Alda code formatting improvement: instrument calls now go on a new
  line.

* Added a ton of docstrings! I'm not 100% sure the formatting is correct, so
  some parts might end up looking wonky. I don't really have a good way to test
  how the docs will look before I deploy them, and the docs are tied to
  releases, so I guess I'll just have to push additional releases if anything
  documentation-related needs fixing. ¯\\\_(ツ)\_/¯

## 0.1.4 (2018-12-22)

No change; bumped version to update documentation via cljdoc.

## 0.1.3 (2018-12-21)

No change; bumped version while tinkering with cljdoc.

## 0.1.2 (2018-12-13)

* Bugfix: `(octave :up)` generates `>`, not `(octave :up)`

## 0.1.1 (2018-12-11)

No change; I had to bump the version while tinkering with deployment setup.

## 0.1.0 (2018-12-11)

Initial release. I think things are relatively stable, but report any issues you
may run into!

[wcerfgba]: https://github.com/wcerfgba
[getting-started]: https://cljdoc.org/d/io.djy/alda-clj/CURRENT/doc/getting-started
[alda-2-syntax-changed]: https://github.com/alda-lang/alda/blob/master/doc/alda-2-migration-guide.md#attribute-syntax-has-changed-in-some-cases


dev/demos.clj
(ns demos
  (:require [alda.core :refer :all]))

;;; basic usage

(comment
  ;; shells out to the alda CLI
  (alda "version")
  (alda "status")

  ;; Forget about current tempo, octave, instruments, etc.
  ;; Start from a clean slate.
  (clear-history!)

  ;; stop playback
  (stop!)

  ;; shorthand for "alda play -c 'piano: o4 c1 / e- / g'"
  (play!
    (part "piano")
    (octave 4)
    (chord (note (pitch :c) (note-length 1))
           (note (pitch :e :flat))
           (note (pitch :g))))

  ;; working with sequences of musical events (notes, rests, etc.)
  (play!
    (part "piano")
    (set-note-length 16)
    (interpose (pause)
               (for [letter [:c :d :e :f :g]]
                 (note (pitch letter)))))

  ;; generate a sequence of random notes
  (play!
    (part "piano")
    (for [n (repeatedly 8 #(+ 100 (rand-int 800)))]
      (note (pitch (rand-nth [:c :e :g :a :b]))
            (ms n)))))

















;;; charge!

(comment
  (clear-history!)

  (def riff "c8 < g a b >")

  (play!
    (part "organ")
    (quant 70) (octave 3)

    (for [[semitones bpm]
          (->> [0 70]
               (iterate (fn [[s b]] [(inc s) (+ b 20)]))
               (take 25))]
      [(transpose semitones)
       (tempo bpm)
       riff])

    (note (pitch :c) (ms 5000))))






















;;; entropy

(comment
  (clear-history!)
  (stop!)

  (def REST-RATE 0.15)
  (def MS-LOWER 30)
  (def MS-UPPER 3000)
  (def MAX-OCTAVE 8)

  (defn random-note
    "Returns a random note in a random octave with a random duration in
     milliseconds.

     May randomly return a rest with a random duration in milliseconds, instead."
    []
    (let [ms (ms (rand-nth (range MS-LOWER MS-UPPER)))]
      (if (< (rand) REST-RATE)
        (pause (duration ms))
        (let [o (rand-int (inc MAX-OCTAVE))
              n [(keyword (str (rand-nth "abcdefg")))
                 (rand-nth [:sharp :flat :natural])]]
          [(octave o)
           (note (apply pitch n) (duration ms))]))))

  (play!
    (part "midi-electric-piano-1")
    (panning 25)
    (repeatedly 50 random-note)

    (part "midi-timpani")
    (panning 50)
    (repeatedly 50 random-note)

    (part "midi-celesta")
    (panning 75)
    (repeatedly 50 random-note)))

























;;; algorithmic bongos

(comment
  (defn random-bongos
    [ticks vol-min vol-max]
    [(octave 4)
     (set-note-length 16) ; 16th notes
     (let [notes     (-> #{[:c] [:c :sharp] [:d] [:d :sharp] [:e]}
                         (->> (map #(note (apply pitch %))))
                         (conj (pause)))
           rand-note #(rand-nth notes)
           rand-vol  #(vol (+ vol-min (rand-int (- vol-max vol-min))))
           tick      #(vector (rand-vol) (rand-note))]
       (take ticks (repeatedly tick)))])

  (play!
    (part "midi-percussion")

    ;; bongos
    (voice 1)
    (random-bongos (* 8 4 4) 25 75)

    ;; hi-hat to keep time
    (voice 2)
    (octave 2)
    (vol 50)
    (set-note-length 4)
    (repeat (* 8 4) (note (pitch :g :sharp)))

    (end-voices)

    ;; ding
    (vol 50) "o5 a1"))



src/alda/shell.clj
(ns ^:no-doc alda.shell
  (:require [me.raynes.conch.low-level :as conch])
  (:import [java.io ByteArrayOutputStream PrintWriter Writer]))

(defn multi-writer-with-autoflush
  "Given a seq of Writers, returns a single Writer that writes the same output
   to all of them. Output is flushed automatically on every write.

   Adapted from: https://stackoverflow.com/a/7987606/2338327"
  [writers]
  ;; I had to comment some of these methods out because Clojure doesn't let you
  ;; do multiple overloads of the same arity. This isn't great, but I guess it's
  ;; alright because you can still accomplish whatever you need to do by using
  ;; the overloads I kept.
  (proxy [Writer] []
    ;; (append [^char c]
    ;;   (run! #(.append % c)))
    (append
      ([^CharSequence csq]
       (run! #(.append % csq)))
      ([^CharSequence csq start end]
       (run! #(.append % csq start end))))
    (close []
      (run! #(.close %) writers))
    (flush []
      (run! #(.flush %) writers))
    (nullWriter []
      (throw (ex-info "Method not supported" {:method "nullWriter"})))
    (write
      ([cbuf]
       (run! #(.write % cbuf) writers)
       (run! #(.flush %) writers))
      ([^chars cbuf offset length]
       (run! #(.write % cbuf offset length) writers)
       (run! #(.flush %) writers)))))
      ;; ([^char c]
      ;;  (run! #(.write % c) writers)))))
      ;; ([^String s]
      ;;  (run! #(.write % s) writers)))))
      ;; ([^String s offset length]
      ;;  (run! #(.write % s offset length) writers)))))

(defn sh
  "Starts a process via me.raynes.conch.low-level/proc.

   Shell output is streamed to stdout and stderr as it is produced
   (implementation adapted from boot.util/sh).

   Like clojure.java.shell/sh, waits for the process to exit and then returns a
   map containing:
     :exit  exit code      (int)
     :out   stdout output  (string)
     :err   stderr output  (string)"
  [cmd & args]
  (let [proc                (apply conch/proc cmd args)
        [out-copy err-copy] (repeatedly 2 #(ByteArrayOutputStream.))
        out-writer          (multi-writer-with-autoflush
                              [*out* (PrintWriter. out-copy)])
        err-writer          (multi-writer-with-autoflush
                              [*err* (PrintWriter. err-copy)])
        out-future          (future (conch/stream-to proc :out out-writer))
        err-future          (future (conch/stream-to proc :err err-writer))
        exit-code           (.waitFor (:process proc))]
    ;; Ensure that all of the output is written to `out-copy` and `err-copy`
    ;; before we return it as a string.
    @out-future
    @err-future
    {:out  (str out-copy)
     :err  (str err-copy)
     :exit exit-code}))


.clj-kondo/hooks.clj
(ns hooks
  (:require [clj-kondo.hooks-api :as api]))

(defn- assert-symbol!
  "Checks whether a `node` is a symbol.

   If it isn't, registers a finding and also returns it."
  [node]
  (when-not (symbol? (api/sexpr node))
    (let [finding (merge
                    {:type    :type-mismatch
                     :message "Argument must be a symbol."}
                    (meta node))]
      (api/reg-finding! finding)
      finding)))

(defn- assert-arity!
  "Checks the number of arguments in a `list-node` against the expected `arity`.

   If there is an arity mismatch, registers a finding and also returns it."
  [arity list-node]
  (let [args       (rest (:children list-node))
        args-count (count args)]
    (when-not (= arity args-count)
      (let [finding (merge
                      {:type    :invalid-arity
                       :message (format "Expected %d arg(s), but got %d"
                                        arity
                                        (count args))}
                      (meta list-node))]
        (api/reg-finding! finding)
        finding))))

(defn lisp-builtin
  "Defines a alda-clj function corresponding to a built-in alda-lisp function.

   Example usage:

   (lisp-builtin set-duration)"
  [{:keys [node]}]
  (let [token  (second (:children node))
        errors (remove nil? [(assert-arity! 1 node)
                             (assert-symbol! token)])]
    (when (empty? errors)
      {:node
       (let [sym (api/sexpr token)]
         (api/list-node
           [(api/token-node 'defn)
            (api/token-node sym)
            (api/token-node
              (format "Emits inline Lisp code `(%s ...)" sym))
            (api/vector-node
              [(api/token-node '&)
               (api/token-node 'args)])
            (api/list-node
              [(api/token-node 'alda.core/map->Sexp)
               (api/map-node
                 [(api/keyword-node :form)
                  (api/list-node
                    [(api/token-node 'list*)
                     (api/list-node
                       [(api/token-node 'quote)
                        (api/token-node sym)])
                     (api/token-node 'args)])])])]))})))

(defn lisp-builtins
  "Defines alda-clj functions corresponding to built-in alda-lisp functions.

   Example usage:

   (lisp-builtins
     set-duration set-note-length)"
  [{:keys [node]}]
  (let [tokens (rest (:children node))
        errors (->> tokens
                    (map assert-symbol!)
                    (remove nil?))]
    (when (empty? errors)
      {:node
       (api/list-node
         (cons
           (api/token-node 'do)
           (for [token tokens
                 :let [sym (api/sexpr token)]]
             (api/list-node
               [(api/token-node 'alda.core/lisp-builtin)
                (api/token-node sym)]))))})))

(defn lisp-builtin-attributes
  "Defines alda-clj functions corresponding to built-in alda-lisp attributes.

   Example usage:

   (lisp-builtin-attributes
     tempo quant volume panning)"
  [{:keys [node]}]
  (let [tokens (rest (:children node))
        errors (->> tokens
                    (map assert-symbol!)
                    (remove nil?))]
    (when (empty? errors)
      {:node
       (api/list-node
         (cons
           (api/token-node 'do)
           (for [token tokens
                 :let [sym (api/sexpr token)]]
             (api/list-node
               [(api/token-node 'alda.core/lisp-builtins)
                (api/token-node sym)
                (api/token-node (symbol (str sym "!")))]))))})))

(defmacro ^:private lisp-builtin-impl
  [sym]
  `(defn ~sym
     ~(format "Emits inline Lisp code `(%s ...)`" sym)
     [& ~'args]
     ;; These don't seem to work, but it doesn't really matter that this macro
     ;; returns _exactly_ what the real one in alda.core does; the important
     ;; thing is just that we `def` something.
     ;; (map->Sexp {:form (list* '~sym ~'args)})
     ;; (alda.core/map->Sexp {:form (list* '~sym ~'args)})
     ;;
     ;; ...These don't seem to work either, though:
     ;; {:form (list* '~sym ~'args)}))
     "placeholder value"))

(defmacro ^:private lisp-builtins-impl
  [& syms]
  (cons 'do
        (for [sym syms]
          `(lisp-builtin ~sym))))

(defmacro ^:private lisp-builtin-attributes-impl
  [& syms]
  (cons 'do
        (for [sym syms]
          `(lisp-builtins ~sym ~(symbol (str sym \!))))))

(defn lisp-builtin2
  [{:keys [node]}]
  {:node (api/macroexpand lisp-builtin-impl node)})

(defn lisp-builtins2
  [{:keys [node]}]
  {:node (api/macroexpand lisp-builtins-impl node)})

(defn lisp-builtin-attributes2
  [{:keys [node]}]
  {:node (api/macroexpand lisp-builtin-attributes-impl node)})


dev/scratch.clj
(ns scratch
  (:require [alda.core :refer :all] :reload-all))

(comment
  (stop!)
  (connect!)
  (disconnect!)
  (println (score-text))
  (new-score!)
  (alda "doctor")
  (alda "--help")
  (play! (part "piano:"))
  (play! (part "bassoon/trumpet \"bars\""))
  (play! (part {:names ["violin" "viola"] :nickname "strings"}))
  (play! (repeat 8 "o4 g8 r"))
  (part "bassoon/trumpet 'bars'")
  (part {:names ["violin" "viola"] :nickname "strings"})
  (play! (note (pitch :c :sharp :sharp :flat) (note-length 8)))
  (note (pitch :c :sharp :sharp :flat) (note-length 8))

  (play!
    (part "piano")
    (octave 2)
    (note (pitch :c) (note-length 2))
    (part "accordion")
    (octave 3)
    (note (pitch :c) (note-length 2))
    (octave! :up)
    (part "piano/accordion")
    (note (pitch :c)))

  (play!
    (part "piano")
    (octave 4)
    (chord (note (pitch :c))
           (note (pitch :e :flat) nil :slur)
           (note (pitch :g) (note-length 2) :slur))
    (octave :up)
    (chord (note (pitch :c))
           (note (pitch :e :flat) nil :slur)
           (note (pitch :g) (note-length 1) :slur))
    (octave 'up)
    (chord (note (pitch :c))
           (note (pitch :e :flat) nil :slur)
           (note (pitch :g) (note-length 1) :slur)))

  (play!
    (part "piano")
    (interpose (pause (duration (note-length 16)))
               (for [letter [:c :d :e :f :g]]
                 (note (pitch letter) (note-length 16)))))

  (play!
    (part "guitar")
    (octave 4)
    (voice 1)
    (for [letter [:c :d :e :f]]
      (note (pitch letter) (note-length 8)))
    (voice 2)
    (for [letter [:e :f :g :a]]
      (note (pitch letter) (note-length 8))))

  (play!
    (part "trombone")
    (octave 3)
    (for [letter [:c :d]]
      (note (pitch letter) (note-length 8)))
    (cram (note-length 4)
      (for [letter [:c :d :e]]
        (note (pitch letter))))
    (note (pitch :f) (note-length 2 {:dots 1})))

  (play!
    (part "piano")
    (set-variable "foo"
                  (note (pitch :c))
                  (note (pitch :d))
                  (note (pitch :c)))
    (get-variable "foo")
    (cram (note-length 4)
      (for [letter [:c :d :e]]
        (note (pitch letter))))
    (note (pitch :f) (note-length 2 {:dots 1})))

  (play!
    (part "piano")
    (for [tempo (repeatedly 4 #(+ 60 (rand-int 200)))]
      [(list 'tempo tempo)
       (for [letter [:c :d :e :f :g]]
         (note (pitch letter) (note-length 8)))]))

  (play!
    '(tempo! 150)
    (part "piano")
    '(pan 0)
    (for [letter [:c :d :e :f :g :c :d :e :f :g]]
      (note (pitch letter) (note-length 8)))
    "\n"
    '(pan 100)
    (for [letter [:c :d :e :f :g :c :d :e :f :g]]
      (note (pitch letter) (note-length 8))))

  (play!
    '(tempo! 250)
    (part "piano")
    (for [pan-value (range 101)]
      [(panning pan-value)
       (note (pitch (rand-nth [:c :d :e :f :g :a :b]))
             (note-length 16))
       "\n"]))

  (play!
    (part "piano")
    (for [n (repeatedly 8 #(+ 100 (rand-int 1500)))]
      (note (pitch :c) (ms n))))

  (defn random-note
    []
    (let [letter     (rand-nth [:c :d :e :f :g :a :b :c])
          accidental (rand-nth [:sharp :flat nil])]
      (note (if accidental
              (pitch letter accidental)
              (pitch letter)))))

  (defn random-chord
    []
    (apply chord (repeatedly (+ 2 (rand-int 4)) random-note)))

  (->str (random-chord))

  (play!
    (part "piano")
    (set-note-length 1)
    (random-chord)
    (random-chord)
    (random-chord))

  (play!
    '(key-sig! '(a major))
    (part "piano")
    (note (pitch :c) (note-length 8))
    (note (pitch :d))
    (note (pitch :e))
    (note (pitch :f))
    (note (pitch :g))))


